python/ray/data/impl/plan.py

```diff
@@ -376,6 +376,10 @@ def is_read_stage(self) -> bool:
             self.has_lazy_input()
             and not self._stages_before_snapshot
             and not self._stages_after_snapshot
+            and (
+                not self._snapshot_blocks
+                or isinstance(self._snapshot_blocks, LazyBlockList)
+            )
```

Comment:
Ah, so I created this `is_read_stage()` method back when there was an explicit pre-optimization `CreationStage` for the read stage, and when I refactored everything to keep the read stage as a dynamic `OneToOneStage` that is created during optimization, this method never got updated.

Good catch and fix!